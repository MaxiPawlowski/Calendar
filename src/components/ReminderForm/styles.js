import { createStyles } from '@material-ui/core';

export default () => 
  createStyles({
    root:{
      display: 'flex',
      flexDirection: 'column'
    },
    geoSuggest:{
      fontSize: '1rem',
      fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif',
      lineHeight: '1.1875em',
    },
    geoInput:{
      borderBottom: '1px solid rgba(0, 0, 0, 0.42)'
    },
    suggestItem:{
      fontSize: '1rem',
      padding: '.5em .65em',
      cursor: 'pointer',
      listStyle: 'none',
    },
    suggestItemActive:{
      background: '#267dc0',
      color: '#fff',
    },
    suggestList:{
      padding: 0,
      borderBottom: '1px solid rgba(0, 0, 0, 0.42)',
      borderLeft: '1px solid rgba(0, 0, 0, 0.42)',
      borderRight: '1px solid rgba(0, 0, 0, 0.42)',
      borderBottomLeftRadius: '4px',
      borderBottomRightRadius: '4px',
    },
    suggestListHidden:{
      border:'none'
    },
  });
