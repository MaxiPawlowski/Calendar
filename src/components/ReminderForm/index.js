import React, { useState, useRef, useEffect } from "react";
import TextField from "@material-ui/core/TextField";
import { Button } from "@material-ui/core";
import { withStyles } from '@material-ui/styles';
import PropTypes from 'prop-types';
import DateFnsUtils from '@date-io/date-fns';
import { MuiPickersUtilsProvider, DateTimePicker } from '@material-ui/pickers';
import Geosuggest from 'react-geosuggest';
import styles from "./styles";

const randomHexColor = () => '#' + (Math.random() * 0xFFFFFF << 0).toString(16)

const ReminderForm = ({ classes, prevValues, handleSave }) => {
  const geoInput = useRef(null);
  const [values, setValues] = useState({
    title: '',
    color: randomHexColor(),
    date: new Date(Date.now()),
    city: ''
  });
  const handleChange = name => event => {
    setValues({
      ...values,
      [name]: name === 'title' ? event.target.value.substring(0, 30) : event.target.value
    });
  };
  const handleSubmit = event => {
    event.preventDefault();
    handleSave(values);
    if (geoInput.current && geoInput.current.clear){
      geoInput.current.clear()
    }
    setValues({
      title: '',
      color: randomHexColor(),
      date: new Date(Date.now()),
      city: ''
    });
  }
  
  useEffect(() => {
    if (prevValues){
      setValues({ ...prevValues})
    }
  }, [prevValues])
  
  return (
    <form className={classes.root} onSubmit={handleSubmit}>
      <TextField
        id="title"
        label="Title"
        value={values.title}
        onChange={handleChange('title')}
        margin="normal"
        required
      />

      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <DateTimePicker
          required
          value={values.date}
          onChange={(date) => handleChange('date')({target:{value:date}})}
        />
      </MuiPickersUtilsProvider>

      <TextField
        id="color"
        label="Color"
        type="color"
        value={values.color}
        onChange={handleChange('color')}
        margin="normal"
        InputLabelProps={{
          shrink: true,
        }}
      />

      <Geosuggest
        inputClassName={classes.geoInput + ' MuiInputBase-input MuiInput-input'}
        suggestItemClassName={classes.suggestItem}
        suggestItemActiveClassName={classes.suggestItemActive}
        suggestsClassName={classes.suggestList}
        suggestsHiddenClassName={classes.suggestListHidden}
        className={classes.geoSuggest}
        initialValue={values.city ? values.city.description: undefined}
        ref={geoInput}
        placeholder="City"
        onSuggestSelect={(params) => {          
          if (params){
            handleChange('city')({ 
              target: {
                value: {
                  location: params.location,
                  description: params.description,
                } 
              } 
            })
          }
        }}
      />

      <Button variant="contained" color="primary" type="submit">
        Submit
      </Button>
    </form>
  );
};

ReminderForm.propTypes = {
  classes: PropTypes.shape({
    root: PropTypes.string.isRequired,
    geoSuggest: PropTypes.string.isRequired,
    geoInput: PropTypes.string.isRequired,
    suggestItem: PropTypes.string.isRequired,
    suggestItemActive: PropTypes.string.isRequired,
    suggestList: PropTypes.string.isRequired,
    suggestListHidden: PropTypes.string.isRequired,
  }).isRequired,

  prevValues: PropTypes.oneOfType([
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired,
      date: PropTypes.instanceOf(Date).isRequired,
      id: PropTypes.number,
      city: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          description: PropTypes.string.isRequired,
          location: PropTypes.shape({
            lat: PropTypes.number.isRequired,
            lng: PropTypes.number.isRequired,
          })
        }),
      ]),
    }),
    PropTypes.bool,
  ]),
  handleSave: PropTypes.func.isRequired
};

ReminderForm.defaultProps = {
  prevValues: false,
}

export default withStyles(styles)(ReminderForm);
