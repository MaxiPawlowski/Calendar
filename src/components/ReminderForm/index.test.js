import React from 'react';
import { shallow, mount } from 'enzyme';
import ReminderForm from "./index";

describe('ReminderForm', () => {
  const mockedDate = 1487076708000;
  let wrapper;
  
  describe('when rendered', () => {
    beforeAll(() => {
      wrapper = shallow(<ReminderForm handleSave={() => {}} />);
    });

    it('should render correctly', () => {
      expect(wrapper.dive()).toMatchSnapshot();
    });
  });

  describe('when submitting', () => {
    const handleSave = jest.fn();
    
    beforeAll(() => {
      wrapper = shallow(<ReminderForm handleSave={handleSave} />).dive();
    });
  
    it('should trigger handleSave from props with the input (default) values', () => {
      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(handleSave).toHaveBeenCalledWith({
        title: '',
        color: '#7fffff',
        date: new Date(mockedDate),
        city: ''
      });
    });
  });

  describe('when submitting with previous values', () => {
    const handleSave = jest.fn();
    
    beforeAll(() => {
      wrapper = mount(
        <ReminderForm
          handleSave={handleSave}
          prevValues={{
            title: 'test title',
            color: '#FFFFFF',
            id: Date.now(),
            date: new Date(mockedDate),
            city: 'test city'
          }}
        />
      );
    });

    it('should trigger handleSave from props with the input values', () => {
      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(handleSave).toHaveBeenCalledWith({
        title: 'test title',
        color: '#FFFFFF',
        id: Date.now(),
        date: new Date(mockedDate),
        city: 'test city'
      });
    });
  });
});
