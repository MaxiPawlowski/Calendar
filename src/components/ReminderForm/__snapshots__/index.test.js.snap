// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderForm when rendered should render correctly 1`] = `
<form
  className="ReminderForm-root-1"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(TextField))
    id="title"
    label="Title"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <MuiPickersUtilsProvider
    utils={[Function]}
  >
    <PickerWithState
      allowKeyboardControl={true}
      ampm={true}
      invalidDateMessage="Invalid Date Format"
      maxDate={2100-01-01T00:00:00.000Z}
      maxDateMessage="Date should not be after maximal date"
      minDate={1900-01-01T00:00:00.000Z}
      minDateMessage="Date should not be before minimal date"
      onChange={[Function]}
      openTo="date"
      orientation="portrait"
      required={true}
      showTabs={true}
      value={2017-02-14T12:51:48.000Z}
      views={
        Array [
          "year",
          "date",
          "hours",
          "minutes",
        ]
      }
      wider={true}
    />
  </MuiPickersUtilsProvider>
  <WithStyles(ForwardRef(TextField))
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    id="color"
    label="Color"
    margin="normal"
    onChange={[Function]}
    type="color"
    value="#7fffff"
  />
  <default_1$1
    autoActivateFirstSuggest={false}
    className="ReminderForm-geoSuggest-2"
    disabled={false}
    fixtures={Array []}
    getSuggestLabel={[Function]}
    highlightMatch={true}
    ignoreEnter={false}
    ignoreTab={false}
    initialValue=""
    inputClassName="ReminderForm-geoInput-3 MuiInputBase-input MuiInput-input"
    maxFixtures={10}
    minLength={1}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onSuggestSelect={[Function]}
    placeholder="City"
    queryDelay={250}
    skipSuggest={[Function]}
    style={Object {}}
    suggestItemActiveClassName="ReminderForm-suggestItemActive-5"
    suggestItemClassName="ReminderForm-suggestItem-4"
    suggestsClassName="ReminderForm-suggestList-6"
    suggestsHiddenClassName="ReminderForm-suggestListHidden-7"
  />
  <WithStyles(ForwardRef(Button))
    color="primary"
    type="submit"
    variant="contained"
  >
    Submit
  </WithStyles(ForwardRef(Button))>
</form>
`;
